scylla/src/transport/locator/tablets.rs

```diff
@@ -0,0 +1,630 @@
+use lazy_static::lazy_static;
+use scylla_cql::cql_to_rust::FromCqlVal;
+use scylla_cql::frame::frame_errors::ParseError;
+use scylla_cql::frame::response::result::{deser_cql_value, ColumnType, TableSpec};
+use thiserror::Error;
+use tracing::warn;
+use uuid::Uuid;
+
+use crate::routing::{Shard, Token};
+use crate::transport::Node;
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Error, Debug)]
+pub(crate) enum TabletParsingError {
+    #[error(transparent)]
+    Parse(#[from] ParseError),
+    #[error("Shard id for tablet exceeds u32 range: {0}")]
+    ShardNum(i32),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct RawTabletReplicas {
+    replicas: Vec<(Uuid, Shard)>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct RawTablet {
+    first_token: Token,
+    last_token: Token,
+    replicas: RawTabletReplicas,
+}
+
+type RawTabletPayload = (i64, i64, Vec<(Uuid, i32)>);
+
+lazy_static! {
+    static ref RAW_TABLETS_CQL_TYPE: ColumnType = ColumnType::Tuple(vec![
+        ColumnType::BigInt,
+        ColumnType::BigInt,
+        ColumnType::List(Box::new(ColumnType::Tuple(vec![
+            ColumnType::Uuid,
+            ColumnType::Int,
+        ]))),
+    ]);
+}
+
+const CUSTOM_PAYLOAD_TABLETS_V1_KEY: &str = "tablets-routing-v1";
+
+impl RawTablet {
+    pub(crate) fn from_custom_payload(
+        payload: &HashMap<String, Vec<u8>>,
+    ) -> Option<Result<RawTablet, TabletParsingError>> {
+        let payload = payload.get(CUSTOM_PAYLOAD_TABLETS_V1_KEY)?;
+        let cql_value = match deser_cql_value(&RAW_TABLETS_CQL_TYPE, &mut payload.as_slice()) {
+            Ok(r) => r,
+            Err(e) => return Some(Err(e.into())),
+        };
+
+        // This could only fail if the type was wrong, but we do pass correct type
+        // to `deser_cql_value`.
+        let result: RawTabletPayload = FromCqlVal::from_cql(cql_value).unwrap();
+
+        let replicas = match result
+            .2
+            .into_iter()
+            .map(|(uuid, shard_num)| match shard_num.try_into() {
+                Ok(s) => Ok((uuid, s)),
+                Err(_) => Err(shard_num),
+            })
+            .collect::<Result<Vec<(Uuid, u32)>, _>>()
+        {
+            Ok(r) => r,
+            Err(shard_num) => return Some(Err(TabletParsingError::ShardNum(shard_num))),
+        };
```

Comment:
I remember reading about optimization in `Vec` that prevents reallocation in case where result and source have the same size - but I'm not sure it will be applied there (and it is an optimization, so it is not guaranteed).

Reply:
Vec docs mention this optimization (with a note that it is not guaranteed to be stable, of course): https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-FromIterator%3CT%3E-for-Vec%3CT%3E
Not sure if this will be applied here as we use Result, but maybe it will?

Reply:
Anyway, I don't see any way to avoid this vector now, so marking it as resolved.